tasks/test_results_finisher.py

```diff
@@ -0,0 +1,153 @@
+from tasks.notify import notify_task_name
+
+log = logging.getLogger(__name__)
+
+test_results_finisher_task_name = "app.tasks.test_results.TestResultsFinisherTask"
+
+
+class TestResultsFinisherTask(BaseCodecovTask, name=test_results_finisher_task_name):
+    async def run_async(
+        self,
+        db_session,
+        chord_result: Dict[str, Any],
+        *,
+        repoid: int,
+        commitid: str,
+        commit_yaml: dict,
+        **kwargs,
+    ):
+        repoid = int(repoid)
+        commit_yaml = UserYaml.from_dict(commit_yaml)
+
+        log.info(
+            "Starting test results finisher task",
+            extra=dict(
+                repoid=repoid,
+                commit=commitid,
+                commit_yaml=commit_yaml,
+            ),
+        )
+
+        lock_manager = LockManager(
+            repoid=repoid,
+            commitid=commitid,
+            report_type=ReportType.COVERAGE,
+            lock_timeout=max(80, self.hard_time_limit_task),
+        )
+
+        try:
+            # this needs to be the coverage notification lock
+            # since both tests post/edit the same comment
+            with lock_manager.locked(
+                LockType.NOTIFICATION,
+                retry_num=self.request.retries,
+            ):
+                return await self.process_async_within_lock(
+                    db_session=db_session,
+                    repoid=repoid,
+                    commitid=commitid,
+                    commit_yaml=commit_yaml,
+                    previous_result=chord_result,
+                    **kwargs,
+                )
+        except LockRetry as retry:
+            self.retry(max_retries=5, countdown=retry.countdown)
+
+    async def process_async_within_lock(
+        self,
+        *,
+        db_session,
+        repoid: int,
+        commitid: str,
+        commit_yaml: UserYaml,
+        previous_result: Dict[str, Any],
+        **kwargs,
+    ):
+        log.info(
+            "Running test results finishers",
+            extra=dict(
+                repoid=repoid,
+                commit=commitid,
+                commit_yaml=commit_yaml,
+                parent_task=self.request.parent_id,
+            ),
+        )
+
+        commit: Commit = (
+            db_session.query(Commit).filter_by(repoid=repoid, commitid=commitid).first()
+        )
+        assert commit, "commit not found"
+
+        notify = True
+
+        if self.check_if_no_success(previous_result):
+            # every processor errored, nothing to notify on
+            return {"notify_attempted": False, "notify_succeeded": False}
+
+        test_instances = (
+            db_session.query(TestInstance)
+            .join(Upload)
+            .join(CommitReport)
+            .join(Commit)
+            .filter(Commit.id_ == commit.id_)
+            .order_by(TestInstance.test_id)
+            .order_by(desc(Upload.created_at))
+            .distinct(TestInstance.test_id)
+            .all()
```

Comment:
consider moving this query to a helper function on `TestInstance` or something since you use it in more than one place so that the logic definitely stays the same